[6.0] Note Custom Field Plugin

TLWebdesign opened the pull request:
I've created a Note Custom Field. Based on the Joomla's default Note Form Field Type. This allows users for example to place notes in between custom fields to guide users.

This note only displays in the backend of the website just like the default Note Form Field. It could be made available in front-end if people would desire this. 

### Summary of Changes
Notes Custom Field added


### Testing Instructions
You can discover the extension and then use it in custom fields. It's called note.
Discover it, publish the extension.
Then go to Article -> Custom Fields. Create a new Field of type "Note"

Check that the note is visible in the backend article edit page. 

Set option Display Front-end to yes. Then check that the option is also displayed in the front-end article view. 

### Actual result BEFORE applying this Pull Request

No Note Custom Field available.

### Expected result AFTER applying this Pull Request

Note Custom Field available.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

TLWebdesign committed:
Note custom field first code

TLWebdesign committed:
added sql and updated year in language files

TLWebdesign committed:
Remvoed unnecessary descriptions and cleaned up xml

brianteeman commented on administrator/components/com_admin/sql/updates/mysql/6.0.0-2024-08-11.sql:
```diff
@@ -4,3 +4,6 @@ UPDATE `#__menu`
 WHERE `link` = 'index.php?option=com_content&view=featured'
 AND `client_id` = 1;
+
+INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `locked`, `manifest_cache`, `params`, `custom_data`, `ordering`, `state`) VALUES
```
dont add to existing sql update files. create your own please

TLWebdesign replied:
Ahhh i was not aware of this. Is there a specific naming convention i need to adhere to?

brianteeman replied:
version-year-month-day 
eg 6.0.0-2025-03-27.sql

TLWebdesign replied:
Thanks i've changed it and used the pr creation date.

brianteeman commented on administrator/components/com_admin/sql/updates/postgresql/6.0.0-2024-08-11.sql:
```diff
@@ -4,3 +4,6 @@ UPDATE "#__menu"
 WHERE "link" = 'index.php?option=com_content&view=featured'
 AND "client_id" = 1;
+
+INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "locked", "manifest_cache", "params", "custom_data", "ordering", "state") VALUES
```
dont add to existing sql update files. create your own please

brianteeman commented on administrator/language/en-GB/plg_fields_note.ini:
```diff
@@ -0,0 +1,15 @@
+
+PLG_FIELDS_NOTE="Fields - Note"
+PLG_FIELDS_NOTE_LABEL="Note (%s)"
+PLG_FIELDS_NOTE_PARAMS_CLASS_DESC="The class used on the div of the note"
```
```suggestion
PLG_FIELDS_NOTE_PARAMS_CLASS_DESC="The class used on the div of the note."
```
sentences should end with a period

brianteeman commented on administrator/language/en-GB/plg_fields_note.ini:
```diff
@@ -0,0 +1,15 @@
+PLG_FIELDS_NOTE_PARAMS_CLASS_DESC="The class used on the div of the note"
+PLG_FIELDS_NOTE_PARAMS_CLASS_LABEL="Note Class"
+PLG_FIELDS_NOTE_PARAMS_CLOSE_DESC="This field controls the display of an "x" to close the note. It takes a value of 'true' (for alerts) or the value for the data-dismiss of the Bootstrap close icon."
+PLG_FIELDS_NOTE_PARAMS_CLOSE_LABEL="Show close button"
```
```suggestion
PLG_FIELDS_NOTE_PARAMS_CLOSE_LABEL="Show Close Button"
```

our UI styleguide says that labelsshould be capitalised

TLWebdesign committed:
Update administrator/language/en-GB/plg_fields_note.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

TLWebdesign committed:
Update administrator/language/en-GB/plg_fields_note.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

QuyTon commented on administrator/language/en-GB/plg_fields_note.ini:
```diff
@@ -0,0 +1,15 @@
+PLG_FIELDS_NOTE_LABEL="Note (%s)"
+PLG_FIELDS_NOTE_PARAMS_CLASS_DESC="The class used on the div of the note"
+PLG_FIELDS_NOTE_PARAMS_CLASS_LABEL="Note Class"
+PLG_FIELDS_NOTE_PARAMS_CLOSE_DESC="This field controls the display of an "x" to close the note. It takes a value of 'true' (for alerts) or the value for the data-dismiss of the Bootstrap close icon."
```
```suggestion
PLG_FIELDS_NOTE_PARAMS_CLOSE_DESC="This field controls the display of an \"x\" to close the note. It takes a value of 'true' (for alerts) or the value for the data-dismiss of the Bootstrap close icon."
```

TLWebdesign committed:
Update administrator/language/en-GB/plg_fields_note.ini

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

QuyTon commented on plugins/fields/note/services/provider.php:
```diff
@@ -0,0 +1,46 @@
+        $container->set(
+            PluginInterface::class,
+            function (Container $container) {
+                $plugin     = new Note(
```
```suggestion
                $plugin = new Note(
```

TLWebdesign committed:
Update plugins/fields/note/services/provider.php

Co-authored-by: Quy Ton <quy@nomonkeybiz.com>

TLWebdesign committed:
moved update sql to own file and undid changes to existing sql

QuyTon commented:
`Required` and `Default Value` are not applicable so this will cause confusion.
For `Show Close Button`, I expect the button to not display if set to `false`.

TLWebdesign committed:
removed unused default option

QuyTon commented:
> i'm not exactly sure what you mean with "required". I've removed the empty default value on the close form field xml.

![45233](https://github.com/user-attachments/assets/fc95b386-715e-489f-aa82-48fae906fe1a)

> The documentation says this about the close button: close (optional) a value of 'true' (for alerts) or the value for the data-dismiss of the Bootstrap close icon

I was expecting it to look like this without the close button.

![45233-alert](https://github.com/user-attachments/assets/29779620-175b-467b-8ab6-eddb559b026a)

TLWebdesign commented:
> > i'm not exactly sure what you mean with "required". I've removed the empty default value on the close form field xml.
> 
> ![45233](https://private-user-images.githubusercontent.com/190299371/432631638-fc95b386-715e-489f-aa82-48fae906fe1a.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4OTIyNDQsIm5iZiI6MTc0NDg5MTk0NCwicGF0aCI6Ii8xOTAyOTkzNzEvNDMyNjMxNjM4LWZjOTViMzg2LTcxNWUtNDg5Zi1hYTgyLTQ4ZmFlOTA2ZmUxYS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxN1QxMjEyMjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05Njg2NTI1ZjUwYTc5MDgxMTZlZmUzZDRkYzUwNDE1ZThiZGMxODA4MWI1ZTI2NmM1ODNlNDg1YzdjN2E4MWRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0FQZA7luEk45VCbLKzg1cISXYzg9jObjLxJWe06s3QI)
> 
> > The documentation says this about the close button: close (optional) a value of 'true' (for alerts) or the value for the data-dismiss of the Bootstrap close icon
> 
> I was expecting it to look like this without the close button.
> 
> ![45233-alert](https://private-user-images.githubusercontent.com/190299371/432632737-29779620-175b-467b-8ab6-eddb559b026a.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQ4OTIyNDQsIm5iZiI6MTc0NDg5MTk0NCwicGF0aCI6Ii8xOTAyOTkzNzEvNDMyNjMyNzM3LTI5Nzc5NjIwLTE3NWItNDY3Yi04YWI2LWVkZGI1NTliMDI2YS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxN1QxMjEyMjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xZDZmNjNlMWE3ODE0NzJhNmYzYjk2ODI0MDQwZWZiNzFjZjMyMTk1MjQzOTIxN2Q5NzBiNjlhMDZlNzc4MGZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.wYCcDSdq8S-VxlqXiiqK9q-e_3Ex05S3FFVCsDYVm5g)

Those settings are not added by the plugin. Do you know of a custom field that has these settings removed so i could check that?

exlemor commented:
I have tested this item :white_check_mark: successfully on 84c1b7c9b07a61a15f1c42802f9994cc598ed406<br /><br />Hi Tom, I have successfully tested this.


(1 question, why can't we have this for the User Accounts area as well? :) )<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45233">issues.joomla.org/tracker/joomla-cms/45233</a>.</sub>

exlemor commented:
> > (1 question, why can't we have this for the User Accounts area as well? :) )
> 
> Hi Emmanuel,
> 
> With User Accounts area do you mean the website front-end? Technically there is no reason why we could not allow this. I've created and tested this already but it's not in the PR because it is not a "plain" implementation of the note form field. If maintainers would want this to also be available in front-end of the website then i could add that.

Hi Tom, no I meant backend actually... I know the backend has Users --> User Notes area but that is not referenced in the user's profile itself which is not so convenient, that's why I liked the idea of this being available in the custom fields for Users that would one could have a notes field inside the user's profile... (vs having to go back and forth between a user's profile and the User Notes area).

TLWebdesign committed:
updated SQL install and update.

added plg_fields_note in right alphabetical order and updated the ordering of the plg_fields to match

TLWebdesign committed:
Merge branch '6.0-dev' into NoteCustomField

muhme commented:
I started testing (and I'm not familar with custom fields). I created a Note custom field with 'notes field description' and 'notes field heading', you see the representation in the screenshot. As mentioned before, some settings like 'Default Value' or 'Editable In' are unused. But what hurts me, if I set 'Required' in the custom field, the article is no longer saveable, see error message:

![screen](https://github.com/user-attachments/assets/0a48e08d-9cbc-48c0-91c7-d6a3c2099fed)

Shouldn't we disable the options that can't be used? And maybe a use case would be helpful. I would like to continue to support testing, thx.

TLWebdesign committed:
Merge branch '6.0-dev' into NoteCustomField

TLWebdesign committed:
hide unused fields

i hid required and default_value. Required would give an error when saving article in front-end/ And default_value is ignored.

TLWebdesign committed:
added front-end display option & hide more unused settings

TLWebdesign committed:
Update note.php

TLWebdesign committed:
fix for proper display_frontend check

brianteeman commented:
I'm not sure if note content really should be an editor field with fully wysiwyg or if it should be just a simple textarea. My feeling is that it should be the latter. But if you go for editor then you need to add some filtering for security purposes eg
`filter="\Joomla\CMS\Component\ComponentHelper::filterText"`

TLWebdesign commented:
> I'm not sure if note content really should be an editor field with fully wysiwyg or if it should be just a simple textarea. My feeling is that it should be the latter. But if you go for editor then you need to add some filtering for security purposes eg `filter="\Joomla\CMS\Component\ComponentHelper::filterText"`

shouldn't it just be filter="safehtml" ? The ability for the end user to add some styling and links is why i would like it to be a editor field tho.

brianteeman commented on administrator/language/en-GB/plg_fields_note.ini:
```diff
@@ -0,0 +1,16 @@
+PLG_FIELDS_NOTE_PARAMS_CLOSE_DESC="This field controls the display of an \"x\" to close the note. It takes a value of 'true' (for alerts) or the value for the data-dismiss of the Bootstrap close icon."
+PLG_FIELDS_NOTE_PARAMS_CLOSE_LABEL="Show Close Button"
+PLG_FIELDS_NOTE_PARAMS_DESCRIPTION_LABEL="Note Content"
+PLG_FIELDS_NOTE_PARAMS_DISPLAY_FRONTEND_LABEL="Display in front-end"
```
Please follow the style guide 

Labels should be capitalised and frontend is a single word

TLWebdesign replied:
Ok i changed it to 'Display in Frontend" or should i also capitalise "in" ?

TLWebdesign replied:
ok i do need to also capitalize "In". Figured it out. thanks!

TLWebdesign committed:
filter added to note content and label capitalized.

TLWebdesign committed:
added white line between field definitions

TLWebdesign committed:
updated label to be fully capitalised

TLWebdesign committed:
changed filter type and set editor buttons to false.

exlemor commented:
I have tested this item :white_check_mark: successfully on c29e09a98d8c9a52c34cd6cc0b490dd809104a2b<br /><br />I have successfully tested this :), thanks @TLWebdesign!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45233">issues.joomla.org/tracker/joomla-cms/45233</a>.</sub>

muhme commented:
I have tested this item :white_check_mark: successfully on c29e09a98d8c9a52c34cd6cc0b490dd809104a2b<br /><br />Tested with JBT graft full package
* Enabled 'Log Almost Everything' and 'Debug System'
* ✅ Created a Note with heading and content, is displayed in backend article editing and is not displayed in Frontend
* ✅ Created heading-only note
* ✅ Created content-only note 
* ✅ Created two note fields, shown in backend and if enabled in frontend
* ✅ No entries in `administrator/logs/*`, no additional errors in JS console

🤔 Note: if no heading is given the note field title is used in backend, but in frontend it is empty<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45233">issues.joomla.org/tracker/joomla-cms/45233</a>.</sub>

TLWebdesign committed:
fixed different behaviour in title text between frontend and backend

muhme committed:
Merge branch '6.0-dev' into NoteCustomField

exlemor commented:
I have tested this item :white_check_mark: successfully on 413398221d02254c4d6e00799a7bea137bb4bd2f<br /><br />I have retested this successfully. Thanks @TLWebdesign! <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45233">issues.joomla.org/tracker/joomla-cms/45233</a>.</sub>

richard67 commented:
P.S. If the name of the new update SQL script is not newer than the latest present one when compared with PHP version compare, it will not run when updating e.g. from 6.0.0-alpha3.

richard67 commented on administrator/components/com_admin/sql/updates/mysql/6.0.0-2025-03-27.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `locked`, `manifest_cache`, `params`, `custom_data`, `ordering`, `state`) VALUES
+(0, 'plg_fields_note', 'plugin', 'note', 'fields', 0, 1, 1, 0, 1, '', '{"class":"alert alert-info","heading":"h4"}', '', 17, 0);
```
```suggestion
(0, 'plg_fields_note', 'plugin', 'note', 'fields', 0, 1, 1, 0, 1, '', '{"class":"alert alert-info","heading":"h4"}', '', -1, 0);
```
New plugins should be installed on update with order number -1 so they are shown at the top.

richard67 replied:
And the script has to be renamed, see my other comment in the main thread.

TLWebdesign replied:
ok i did so. thanks :D

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/6.0.0-2025-03-27.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "locked", "manifest_cache", "params", "custom_data", "ordering", "state") VALUES
+(0, 'plg_fields_note', 'plugin', 'note', 'fields', 0, 1, 1, 0, 1, '', '{"class":"alert alert-info","heading":"h4"}', '', 17, 0);
```
```suggestion
(0, 'plg_fields_note', 'plugin', 'note', 'fields', 0, 1, 1, 0, 1, '', '{"class":"alert alert-info","heading":"h4"}', '', -1, 0);
```
New plugins should be installed on update with order number -1 so they are shown at the top.

TLWebdesign committed:
Update administrator/components/com_admin/sql/updates/postgresql/6.0.0-2025-03-27.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

TLWebdesign committed:
changed name of sql

TLWebdesign committed:
Merge branch 'NoteCustomField' of https://github.com/TLWebdesign/joomla-cms into NoteCustomField

richard67 commented on administrator/components/com_admin/sql/updates/mysql/6.0.0-2025-08-16.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO `#__extensions` (`package_id`, `name`, `type`, `element`, `folder`, `client_id`, `enabled`, `access`, `protected`, `locked`, `manifest_cache`, `params`, `custom_data`, `ordering`, `state`) VALUES
+(0, 'plg_fields_note', 'plugin', 'note', 'fields', 0, 1, 1, 0, 1, '', '{"class":"alert alert-info","heading":"h4"}', '', 17, 0);
```
```suggestion
(0, 'plg_fields_note', 'plugin', 'note', 'fields', 0, 1, 1, 0, 1, '', '{"class":"alert alert-info","heading":"h4"}', '', -1, 0);
```
You should not mark review suggestions as resolved if they are not fully resolved.

TLWebdesign committed:
Update 6.0.0-2025-08-16.sql

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 committed:
Merge branch '6.0-dev' into NoteCustomField
